File: 2022-12-12-advent-of-code-day3/day3/main.py
from pathlib import Path

here = Path(__file__).absolute().parent


def compute_priority(item: str) -> int:
    if item.isupper():
        return ord(item) - ord("A") + 27
    else:
        return ord(item) - ord("a") + 1


def compute_total(rucksack: str) -> int:
    fst = set(rucksack[: len(rucksack) // 2])
    snd = set(rucksack[len(rucksack) // 2 :])
    intersection = fst & snd
    if intersection:
        unique_item = (intersection).pop()
        return compute_priority(unique_item)
    else:
        return 0


def compute_total_part1(text: str) -> int:
    return sum(map(compute_total, map(str.strip, text.strip().split("\n"))))


def main() -> int:

    with open(here / "input.txt") as f:
        text = f.read()

    print(f"Part 1:\n{compute_total_part1(text)}")
    # print(f"Part 2:\n{compute_total_part2(text)}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
else:
    import pytest

    EXAMPLE_INPUT = """
    vJrwpWtwJgWrhcsFMMfFFhFp
    jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
    PmmdzqPrVvPwwTWBwg
    wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
    ttgJtRGJQctTZtZT
    CrZsJsPPZsGzwwsLwLmpwMDw
    """

    def test_compute_total_part1():
        assert compute_total_part1(EXAMPLE_INPUT) == 157

    @pytest.mark.parametrize(
        ("rucksack", "expected_priority"),
        (
            ("vJrwpWtwJgWrhcsFMMfFFhFp", 16),
            ("jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL", 38),
            ("PmmdzqPrVvPwwTWBwg", 42),
            ("wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn", 22),
            ("ttgJtRGJQctTZtZT", 20),
            ("CrZsJsPPZsGzwwsLwLmpwMDw", 19),
        ),
    )
    def test_compute_total(rucksack, expected_priority):
        assert compute_total(rucksack) == expected_priority

    @pytest.mark.parametrize(
        ("letter", "expected_priority"), (("a", 1), ("z", 26), ("A", 27), ("Z", 52))
    )
    def test_compute_priority(letter, expected_priority):
        assert compute_priority(letter) == expected_priority

File: 2022-12-12-advent-of-code-day3/day3/src/main.rs
use std::collections::HashSet;
use std::fs;

fn compute_priority(item: &char) -> u32 {
    if item.is_ascii_uppercase() {
        *item as u32 - 65 + 27
    } else {
        *item as u32 - 96
    }
}

fn compute_total(rucksack: &str) -> u32 {
    let size = rucksack.len() / 2;
    let (fst, snd) = rucksack.split_at(size);
    let fst_set: HashSet<char> = fst.chars().collect();
    let snd_set: HashSet<char> = snd.chars().collect();
    let common_item = fst_set.intersection(&snd_set).into_iter().next().unwrap();
    compute_priority(common_item)
}

fn compute_total_part1(text: &str) -> u32 {
    text.trim()
        .split("\n")
        .map(|x| x.trim())
        .map(compute_total)
        .sum()
}

fn main() {
    let file_path = "input.txt";
    let contents = fs::read_to_string(file_path).expect("Should have been able to read the file");
    let total = compute_total_part1(&contents);

    println!("Part 1: {}", total);
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_compute_total_part1() {
        let example_input = "
        vJrwpWtwJgWrhcsFMMfFFhFp
        jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
        PmmdzqPrVvPwwTWBwg
        wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
        ttgJtRGJQctTZtZT
        CrZsJsPPZsGzwwsLwLmpwMDw
        ";
        assert_eq!(compute_total_part1(example_input), 157);
    }
}
